src/main.c

```diff
@@ -453,23 +453,100 @@ static int dev_opt(const char *str) {
     !!strchr(str, '/');
 }
 
+typedef struct {
+  size_t dist;
+  int common_modes;
+  const char *pgmid;
+  const char *desc;
+} pgm_distance;
+
+static int cmp_pgmid(const void *a, const void *b) {
+   const pgm_distance *pa = a, *pb = b;
+
+   int ret = pa->dist - pb->dist;
+   if(ret)
+     return ret;
+   return strcmp(pa->pgmid, pb->pgmid);
+}
+
+static int suggest_programmers(const char *programmer, LISTID programmers) {
+  const int max_distance = 64; // Don't show suggestions if they are way far out
+
+  size_t nid = 0;              // Number of possible programmer ids
+  for(LNODEID ln1 = lfirst(programmers); ln1; ln1 = lnext(ln1)) {
+    PROGRAMMER *pgm = ldata(ln1);
+    if(is_programmer(pgm))
+      for(LNODEID ln2 = lfirst(pgm->id); ln2; ln2 = lnext(ln2))
+        nid++;
+  }
+
+  pgm_distance *d = cfg_malloc(__func__, nid*sizeof*d);
+
+  // Fill d[] struct
+  size_t idx = 0;
+  AVRPART *p = locate_part(part_list, partdesc);
+  for(LNODEID ln1 = lfirst(programmers); ln1; ln1 = lnext(ln1)) {
+    PROGRAMMER *pgm = ldata(ln1);
+    if(!is_programmer(pgm))
+      continue;
+    for(LNODEID ln2 = lfirst(pgm->id); ln2; ln2 = lnext(ln2)) {
+      if(idx < nid) {
+        d[idx].pgmid = ldata(ln2);
+        d[idx].desc = pgm->desc;
+        d[idx].dist = str_weighted_damerau_levenshtein(d[idx].pgmid, programmer);
+        d[idx].common_modes = pgm->prog_modes & (p? p->prog_modes: -1);
+        idx++;
+      }
+    }
+  }
+
+  size_t n = 0, pgmid_maxlen = 0, comp = 0;
+  if(nid) {                     // Sort list so programmers according to string distance
+    qsort(d, nid, sizeof(*d), cmp_pgmid);
+    size_t dst = d[nid > 2? 2: nid-1].dist; // Print at least 3 close suggestions if possible
+    if(dst > max_distance)
+      dst = max_distance;
+    for(; n < nid; n++) {
+      if(d[n].dist > dst)
+        break;
+      size_t len = strlen(d[n].pgmid);
```

Comment:
Small point: maxlen should not be affected by programmers that we are no longer considering to print.